Tasks/FtpUpload/ftpuploadtask.ts

```diff
@@ -0,0 +1,243 @@
+/// <reference path="../../definitions/vsts-task-lib.d.ts" />
+/// <reference path="../../definitions/node.d.ts" />
+/// <reference path="../../definitions/Q.d.ts" />
+
+import path = require('path');
+import tl = require('vsts-task-lib/task');
+import url = require('url');
+import Q = require('q');
+
+import ftputils = require('./ftputils');
+
+var SortedSet = require('collections/sorted-set');
+var Client = require('ftp');
+
+var win = tl.osType().match(/^Win/);
+tl.debug('win: ' + win);
+
+var repoRoot: string = tl.getVariable('build.sourcesDirectory');
+function makeAbsolute(normalizedPath: string): string {
+    tl.debug('makeAbsolute:' + normalizedPath);
+
+    var result = normalizedPath;
+    if (!path.isAbsolute(normalizedPath)) {
+        result = path.join(repoRoot, normalizedPath);
+        tl.debug('Relative file path: ' + normalizedPath + ' resolving to: ' + result);
+    }
+    return result;
+}
+
+// server endpoint
+var serverEndpoint = tl.getInput('serverEndpoint', true);
+var serverEndpointUrl: url.Url = url.parse(tl.getEndpointUrl(serverEndpoint, false));
+
+var serverEndpointAuth = tl.getEndpointAuthorization(serverEndpoint, false);
+var username = serverEndpointAuth['parameters']['username'];
+var password = serverEndpointAuth['parameters']['password'];
+
+// the root location which will be uploaded from
+var rootFolder: string = makeAbsolute(path.normalize(tl.getPathInput('rootFolder', true).trim()));
+if (!tl.exist(rootFolder)) {
+    failTask('The specified root folder does not exist: ' + rootFolder);
```

Comment:
This code is not needed.  The agent will resolve the path input and by you calling true as the second arg above to check if exists, this will never be true.